### PR TITLE
Add ContactModel1D

### PR DESCRIPTION
**Pull request by skleff1994**
Implementation of a contact model 1D that constrains only the translation along z-axis . It allows to apply a normal force while moving tangentially, which is useful for e.g. scraping or sanding tasks, but not allowed by current contact models (2D, 3D, 6D). 

More specifically : 
- implementation of ContactModel1D 
- completion the contacts unittests factory 
- implemention of python binding
- completion of ResidualModelContactForce to allow normal force cost

I'm not sure if things are missing, please let me know !

### CODE REVIEW

**Review by jcarpent — COMMENTED** [review]
All these contacts models are solving very similar problems. The only change lies in the number of authorized dof.
To simplify the task of code maintenance and make the code as generic as possible, I would rather suggest using a selection matrix which only projects the quantities on the constraint manifold, having a single base class.

**Comment by jcarpent** [comment]
In addition, the flow of commits should not contain merged commits, so far and so on, which drastically reduce the overall readability of the history.
It is better to rely on a `git rebase` strategy.

File: include/crocoddyl/multibody/fwd.hpp
Comment (jcarpent) [inline include/crocoddyl/multibody/fwd.hpp:475]: Here ContactUndefined should be the first element of the enum for debugging perspectives.

File: include/crocoddyl/multibody/fwd.hpp
Comment (jcarpent) [inline include/crocoddyl/multibody/fwd.hpp:476]: The same applies here.

**Review by nim65s — APPROVED** [review]
I agree with @jcarpent comments, but I think refactoring `Contact{1,2,3,6}D → ContactND` could be handled in another PR. The same apply for the `ContactUndefined` issue.

Also, the commit history is a bit messy. But as the final diff is pretty simple, we can use the "squash and merge" button, to have only one commit in the end. @skleff1994 : if you want to go ahead and rewrite the history & `push -f`, it might be the occasion to learn and improve that for the next time, but that's not mandatory here IMHO :)

**Comment by skleff1994** [comment]
Thanks for your feedback, I cleaned up the commits history . I'll fix the CI for this PR and open a new PR for contact models unification as discussed

File: bindings/python/crocoddyl/multibody/contacts/contact-1d.cpp
Comment (cmastalli) [inline bindings/python/crocoddyl/multibody/contacts/contact-1d.cpp:20]: Removed these lines.
We used to disable the print of deprecated code.
However, this shouldn't be the case for a new class.

File: bindings/python/crocoddyl/multibody/contacts/contact-1d.cpp
Comment (cmastalli) [inline bindings/python/crocoddyl/multibody/contacts/contact-1d.cpp:92]: Also remove this line

File: bindings/python/crocoddyl/multibody/contacts/contact-1d.cpp
Comment (cmastalli) [inline bindings/python/crocoddyl/multibody/contacts/contact-1d.cpp:87]: Also remove this line.
We don't want to push this deprecated method.
We deprecate a function when it has been previously develop.
This helps users to upgrade users' code.

File: bindings/python/crocoddyl/multibody/contacts/contact-1d.cpp
Comment (cmastalli) [inline bindings/python/crocoddyl/multibody/contacts/contact-1d.cpp:26]: Does constraints the acceleration in z only? right?
Please update the doc appropriately.

File: bindings/python/crocoddyl/multibody/contacts/contact-1d.cpp
Comment (cmastalli) [inline bindings/python/crocoddyl/multibody/contacts/contact-1d.cpp:46]:
```suggestion
          bp::args("self", "state", "id", "nu", "gains"),
```

Please also update the documentation in line 49

File: include/crocoddyl/multibody/contacts/contact-1d.hpp
Comment (cmastalli) [inline include/crocoddyl/multibody/contacts/contact-1d.hpp:110]: Remove this function and its implementation in .hpp file

File: bindings/python/crocoddyl/multibody/contacts/contact-1d.cpp
Comment (cmastalli) [inline bindings/python/crocoddyl/multibody/contacts/contact-1d.cpp:57]: Removed this constructor.
We have deprecated `FrameTranslation` object.
Soon it will be erased in Crocoddyl.

File: include/crocoddyl/multibody/contacts/contact-1d.hpp
Comment (cmastalli) [inline include/crocoddyl/multibody/contacts/contact-1d.hpp:71]: Removed these constructors and their implementations in .hpp file.

File: include/crocoddyl/multibody/contacts/contact-1d.hxx
Comment (cmastalli) [inline include/crocoddyl/multibody/contacts/contact-1d.hxx:45]: Remove these lines as well (as justified above)

File: include/crocoddyl/multibody/contacts/contact-1d.hxx
Comment (cmastalli) [inline include/crocoddyl/multibody/contacts/contact-1d.hxx:59]: It seems this is not compatible with codegen.
It should be

```suggestion
  d->Jc.template row(0) = d->fJf.template row(2);
```

File: include/crocoddyl/multibody/contacts/contact-1d.hxx
Comment (cmastalli) [inline include/crocoddyl/multibody/contacts/contact-1d.hxx:88]: Same here, it is not compatible with codegen.
See below for details on how to fix it.

File: include/crocoddyl/multibody/contacts/contact-1d.hxx
Comment (cmastalli) [inline include/crocoddyl/multibody/contacts/contact-1d.hxx:103]: Same here

File: include/crocoddyl/multibody/contacts/contact-1d.hxx
Comment (cmastalli) [inline include/crocoddyl/multibody/contacts/contact-1d.hxx:145]: Remove these lines

File: include/crocoddyl/multibody/fwd.hpp
Comment (cmastalli) [reply include/crocoddyl/multibody/fwd.hpp:475]: @skleff1994 -- could you please handle it?

File: include/crocoddyl/multibody/residuals/contact-force.hxx
Comment (cmastalli) [inline include/crocoddyl/multibody/residuals/contact-force.hxx:44]: Why are you using `Eigen::Matrix<Scalar, 1, 1>`?

I think it will allocate memory.

File: include/crocoddyl/multibody/residuals/contact-force.hxx
Comment (cmastalli) [inline include/crocoddyl/multibody/residuals/contact-force.hxx:74]: This looks wrong. You should pass the row that you need to constrain.

File: unittest/factory/contact.cpp
Comment (cmastalli) [inline unittest/factory/contact.cpp:67]: Minor detail:

No need to cast 0. as a double, just write `0.`

**Review by cmastalli — CHANGES_REQUESTED** [review]
Apologies for a late review, I appreciate your patience!
There are a few comments that still needs to be handle before merging this request.

File: include/crocoddyl/multibody/residuals/contact-force.hxx
Comment (jcarpent) [reply include/crocoddyl/multibody/residuals/contact-force.hxx:44]: This is on the stack.

File: include/crocoddyl/multibody/fwd.hpp
Comment (skleff1994) [reply include/crocoddyl/multibody/fwd.hpp:475]: Thank you for reminding me!  Yes I will handle it.

**Comment by skleff1994** [comment]
Thanks for the review, and sorry for the delay. I tried to address all the requested changes, and I just fixed the CI. Please let me know if I can do anything else for this PR !

File: bindings/python/crocoddyl/multibody/contacts/contact-1d.cpp
Comment (cmastalli) [inline bindings/python/crocoddyl/multibody/contacts/contact-1d.cpp:11]: We do not need to include any more this header.
Could you remove it?

File: include/crocoddyl/multibody/contacts/contact-1d.hpp
Comment (cmastalli) [inline include/crocoddyl/multibody/contacts/contact-1d.hpp:20]: same here

**Review by cmastalli — COMMENTED** [review]
@skleff1994 apologies for asking some extra (but very tiny) modifications.
Could you please remove the deprecate headers? (see my two comments for more details)

**Review by cmastalli — APPROVED** [review]

**Comment by cmastalli** [comment]
I merge this as I don't want to wait for gitlab-ci issues (it passes the other tests).
@skleff1994 thank you so much for your patience.

**Comment by wxmerkt** [comment]
**After this PR, Crocoddyl fails with clang-10/clang-12 but works with g++-9 under Ubuntu 20.04**

I am getting build failures in the Python bindings from `contact-1d.hxx` using clang-12 on Ubuntu 20.04 (Eigen 3.3.7):
```
FAILED: bindings/python/crocoddyl/CMakeFiles/crocoddyl_pywrap.dir/multibody/contacts/contact-1d.cpp.o 
/usr/bin/clang++-12  -DBOOST_MPL_LIMIT_LIST_SIZE=30 -DBOOST_MPL_LIMIT_VECTOR_SIZE=30 -DPINOCCHIO_WITH_URDFDOM -Dcrocoddyl_pywrap_EXPORTS -I. -Iinclude -I../include -I../bindings -isystem /usr/include/eigen3 -isystem /home/wxm/dev/install_ws/install/include -isystem /usr/include/python3.8 -isystem /usr/local/lib/python3.8/dist-packages/numpy/core/include -pedantic -Wno-long-long -Wall -Wextra -Wcast-align -Wcast-qual -Wformat -Wwrite-strings -Wconversion  -O2 -g -DNDEBUG -fPIC   -Wall -Wpedantic -Wextra -Wno-sign-conversion -Wno-conversion -Wno-deprecated-declarations -MD -MT bindings/python/crocoddyl/CMakeFiles/crocoddyl_pywrap.dir/multibody/contacts/contact-1d.cpp.o -MF bindings/python/crocoddyl/CMakeFiles/crocoddyl_pywrap.dir/multibody/contacts/contact-1d.cpp.o.d -o bindings/python/crocoddyl/CMakeFiles/crocoddyl_pywrap.dir/multibody/contacts/contact-1d.cpp.o -c ../bindings/python/crocoddyl/multibody/contacts/contact-1d.cpp
In file included from ../bindings/python/crocoddyl/multibody/contacts/contact-1d.cpp:10:
In file included from ../include/crocoddyl/multibody/contacts/contact-1d.hpp:206:
../include/crocoddyl/multibody/contacts/contact-1d.hxx:37:18: error: 'row' following the 'template' keyword does not refer to a template
  d->Jc.template row(0) = d->fJf.template row(2);
        ~~~~~~~~ ^
/usr/include/boost/python/object/make_ptr_instance.hpp:61:23: note: in instantiation of member function 'crocoddyl::ContactModel1DTpl<double>::calc' requested here
            type_info(typeid(*x))
                      ^
```

It passed the buildserver which uses g++ by default. I'll add another config with clang to catch this in the future.

Any advice on how to resolve this / which compiler settings did you use to get it to work?

**Comment by jcarpent** [comment]
There is a mistake in the code. The `template` keyword should not be there. clang, as usual, is correct.

**Comment by wxmerkt** [comment]
Thank you for the quick confirmation @jcarpent. I found a similar issue & fix in caffe2 and am currently preparing a PR with a fix

**Pull request merged**
